### ISSUE TITLE
Json API data LIVE

### ISSUE THREAD

**Issue by IVEN2022** [Cuperino/QPrompt#141]
Is there a way to make Qprompt take data directly from live Json api  url feed , I hope this can be possible

**Comment by Cuperino** [comment]
Could you expand more into your use case?

There are plans to include networking features in 2023-2024, such adding a rest API for remote controlling QPrompt and triggering HTTP requests when programmable events take place. But I need to understand your use case better so needs can be appropriately met.

**Comment by IVEN2022** [comment]
well since we use some software to write the news segments on it , a rest Json API feed can contain the data that we need to put on the qprompet , so it only needs to ready the data from the live api feed , so when ever we change the data in the api feed , the qprompet will be auto updated .

**Comment by Cuperino** [comment]
I see. Do you know if the JSON from the news API is pushed into other software's IP addresses, or if QPrompt would have to pull it periodically?

Also, could you attach a sample of said JSON? It's just to have an idea of what values QPrompt would need to read, and how could we make this easy to configure.

**Comment by IVEN2022** [comment]
yes json is already pushed via  ip  no need for Qprompt to pull it  , just need to read it , here is a raw data of the json api

https://pastebin.com/raw/f6MT0Qb4 , if you need more info please let me know

**Comment by Cuperino** [comment]
Thanks!

Copy pasting here in case the pastebin ever goes down:
```
[{"id":1514,"HeadLine":"Germany coup plot: The extremists who tried to topple the state","SubLine":"Three alpaca stared bleakly at us as we rounded what seemed like the umpteenth sharp bend along the rural roads of the east German region of east Thuringia.","Location":"BERLIN","Headline_Image":false},{"id":1510,"HeadLine":"Lockerbie bombing suspect in US custody","SubLine":"A Libyan man accused of making the bomb which destroyed Pan Am flight 103 over Lockerbie 34 years ago is in United States custody, Scottish authorities have said.","Location":"Lockerbie","Headline_Image":false}]
```

How does this sound to you?

- HeadLines could be placed in a feed on the sidebar.
- As new headlines come in, the feed is updated.
- The operator would have the option to be notified through the operating system's notifications.
- The operator can then drag and drop the headline block from the sidebar into the editor for it to be prompted.
- There would also be an option to automatically append the feed's new contents to the end of the current document.

**Comment by IVEN2022** [comment]
well you suggestion is great , since all the documents instead of to be in a file , the full document is brought via the json feed , since in on our news editor software , we can organized what to be sent to the api feed , and also by Order. , when we remove the data from the feed ( data status : OFFAIR ) then the api feed will contains nothing , only data that is STATUS: ON AIR will be sent to the feed , so in short , imaging the full news document is in that feed ,organized and in order , all is needed is that the QP to read the feed lets say each 1 second , the QP will send get request , this way anything is updated or corrected or added to the news feed will be on the QP on the fly

**Comment by IVEN2022** [comment]
oh i hope this is not 2023-2024 project , this is long time

**Comment by Cuperino** [comment]
Software development takes time, and there's a lot of testing involved. I'm also the only developer currently working on QPrompt, and I have a full time job as a software developer, that doesn't involve QPrompt.

I will add this request to a poll that I'm making for Patreon supporters. Patreons determine which features I should work on next. Your support on Patreon won't make a drastic difference in the timeline, but it's still welcome and would give you early access to binaries with the new features as they're developed.

QPrompt's Patreon page: https://www.patreon.com/qpromptapp

**Comment by Cuperino** [comment]
More points for reference:
- Feed configurations are saved so they can be reused in the future.
- There could be more than one feed configuration. Users should be able to add, modify, and delete feed configurations.
- Because feeds could come in different data models, users needs an interface to configure how a feed gets formatted. This would only need to be setup once per feed.
- When feeds are enabled, the most recently used feed would be automatically selected.
- Data is pulled periodically through a GET request.
- The data from the example feed is curated, so it could all go straight to the prompter.

**Comment by videosmith** [comment]
Not as efficient as the suggested api, however as a crude alternative, could Qprompt periodically poll or be triggered to scan a loaded html source document for changes in creation time, then automatically load the changed version in a buffer and swap with current script matching the scroll position (assuming no changes are implemented prior to the current scroll position)?

**Comment by Cuperino** [comment]
Well, as you're probably aware @videosmith, that is already a feature QPrompt has. Local documents are listened for changes, and when another program edits them, they're reloaded automatically. When the document reloads, the prompter's position is maintained.

I don't know if this will work with files being read from a Network Attached Storage... I haven't set one up.

The drawbacks from our current implementation are the following:
- When the document reloads, its undo/redo history gets reset; which can be problematic if two people work on the document at the same time.
- Loading the document locks the GUI thread, meaning animations will stop for a few milliseconds and resume at a later point in time. This can be very noticeable on lengthy scripts, and when the scripts are stored on mechanical hard drives, slow flash drives, and slow SD cards.

QPrompt versions prior to 1.1.2 also have the following issue:
- When a storage medium lies to the operating system about the drive's cache being flushed to disk (a common thing to see with flash drives and SD Cards), this feature would trigger the following bug while saving: #126

**Comment by IVEN2022** [comment]
nice , the milestone is getting faster , i hope it will be available soon

**Comment by Cuperino** [comment]
As you can see, some infrastructure is already in place, but as I mentioned previously, this is likely to not be usable for a year or two.

I'll keep this issue up to date as milestones are accomplished. Once a feature is added, you can build the code to test or use it, or download one of the early access pre-releases Patreons have access to (which I should start posting in February, after the latest bug fix release for 1.1.x is out, which I'm planning to release on Monday.).

**Comment by IVEN2022** [comment]
My god ,   """"""""""" this is likely to not be usable for a year or two"""""""
Year or two . wow this is so much time to update a feature ,""  
maybe not on my time , i will not be on this world in 2 years ,  so yeah .

**Comment by Cuperino** [comment]
I'm very sorry to hear that. I'll see what I can do.

Code contributions are welcome. There's a large backlog of requested features and little time for me to work on them (mostly some weekends).
